### ISSUE TITLE
[Analyzer] false-positive on clang-analyzer-core.uninitialized.Branch on coroutine promise type

### ISSUE THREAD

**Issue by KJTsanaktsidis** [llvm/llvm-project#193397]
It seems like this analysis doesn't correctly understand that `PromiseType` gets default-initialized?

```c++
#include <coroutine>
#include <exception>

namespace repro
{

struct ReturnObject;

struct PromiseType
{
    bool receiver;

    PromiseType() : receiver(false) {}

    ReturnObject get_return_object();

    std::suspend_never initial_suspend() noexcept { return {}; }
    std::suspend_never final_suspend() noexcept { return {}; }
    void unhandled_exception() { std::terminate(); }

    void return_value(int value)
    {
        if (receiver) {}
    }
};

struct ReturnObject
{
    using promise_type = PromiseType;
};

ReturnObject PromiseType::get_return_object()
{
    return {};
}

ReturnObject co_return_after_branch(bool fail)
{
    if (fail)
    {
        co_return 1;
    }
    co_return 0;
}

} // namespace repro

int main()
{
    auto promise = repro::co_return_after_branch(true);
    (void)promise;
    return 0;
}
```

```
%   /nix/store/h3ikd8i3dq1hd7xpzy15kjj7nkwc7x59-clang-tools-22.1.2/bin/clang-tidy \
    clang_analyzer_cobalt_promise_repro.cpp \
    -checks="-*,clang-analyzer-core.uninitialized.Branch" \
    -warnings-as-errors=clang-analyzer-core.uninitialized.Branch \
    -- \
    -std=c++23 \
    -I.
1 warning generated.
clang_analyzer_cobalt_promise_repro.cpp:23:13: error: Branch condition evaluates to a garbage value [clang-analyzer-core.uninitialized.Branch,-warnings-as-errors]
   23 |         if (receiver) {}
      |             ^
clang_analyzer_cobalt_promise_repro.cpp:50:20: note: Calling 'co_return_after_branch'
   50 |     auto promise = repro::co_return_after_branch(true);
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
clang_analyzer_cobalt_promise_repro.cpp:39:9: note: 'fail' is true
   39 |     if (fail)
      |         ^~~~
clang_analyzer_cobalt_promise_repro.cpp:39:5: note: Taking true branch
   39 |     if (fail)
      |     ^
clang_analyzer_cobalt_promise_repro.cpp:41:9: note: Calling 'PromiseType::return_value'
   41 |         co_return 1;
      |         ^~~~~~~~~~~
clang_analyzer_cobalt_promise_repro.cpp:23:13: note: Branch condition evaluates to a garbage value
   23 |         if (receiver) {}
      |             ^~~~~~~~
1 warning treated as error
```

Got this by minimising an example out of a codebase that used `boost::cobalt::promise`, so I've hit this in the real world...

**Comment by NagyDonat** [comment]
Unfortunately the clang static analyzer does not have proper support for coroutiones, e.g. the only reference to `CoreturnStmt` [(source link)](https://github.com/llvm/llvm-project/blob/2e0112ff197d94c4a732fbae61fab9db514c36b3/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp#L1742) is that we abort the analysis if we reach it.  [1]

Nevertheless, the analyzer should handle these in a conservative fashion and should not emit false positives. ~~I'm surprised to see that instead of the above mentioned "silently abort the analysis" we run into this false positive.~~ 

EDIT: I realized that the above mentioned only reference to `CoreturnStmt` is actually dead code: the method `ExprEngine::Visit` is only called for statements in the middle of a `CFGBlock` (so it is never called for `return`, `break`, and e.g. `co_return`). I think the CFG generation logic of clang (which is used by the static analyzer essentially as a library) is aware of the control flow implications of `co_return`, but the static analyzer itself is not aware of coroutines so it doesn't know that in this situation an implicit constructor call happens.

[1] The control flow handling "core" part of the analyzer was written around 2010-12 and since then it "ossified" and became hard to change due to technical debt. Currently there are not enough contributors on the project to implement features like this in the foreseeable futures.

**Comment by steakhal** [comment]
https://godbolt.org/z/Gsre1vWj5
